src/storage.rs

```diff
@@ -0,0 +1,135 @@
+    type Error;
+
+    /// Writes the word to the address
+    fn try_write(&mut self, address: Address<U>, word: Word) -> nb::Result<(), Self::Error>;
+}
+
+/// Read multiple bytes from the device.
+///
+/// Intended to be used for when there is a optimized method of reading multiple bytes.
+///
+/// Iterating over the slice is a valid method to ```impl``` this trait.
+pub trait MultiRead<Word, U> {
+    /// An enumeration of Storage errors
+    type Error;
+
+    /// Reads the words stored at the address to fill the buffer
+    /// ```
+    /// pub fn try_read_slice(
+    ///     &mut self,
+    ///     address: Address,  
+    ///     buf: &mut [Word]
+    /// ) -> nb::Result<Word, Self::Error>
+    ///     let address = address.0 as *const _;
+    ///     unsafe {
+    ///         buf = core::slice::from_raw_parts::<'static, Word>(address,buf.len())
+    ///     }
+    ///     
+    ///     Ok()
+    /// }
+    /// ```
+    fn try_read_slice(
+        &mut self,
+        address: Address<U>,
+        buf: &mut [Word],
+    ) -> nb::Result<(), Self::Error>;
+}
+
+/// Write multiple bytes to the device.
+///
+/// Intended to be used for when there is a optimized method of reading multiple bytes.
+///
+/// Iterating over the slice is a valid method to ```impl``` this trait.
+pub trait MultiWrite<Word, U> {
+    /// An enumeration of Storage errors
+    type Error;
+
+    /// Writes the buffer to the address
+    fn try_write_slice(&mut self, address: Address<U>, buf: &[Word])
+        -> nb::Result<(), Self::Error>;
+}
+
+/// A common interface to erase pages or memory locations.
+///
+/// For Flash storage, the write functions can't set a bit to 1.
+///
+/// For non flash devices, this trait is not required, but it can be used to erase data as recommended by the device (EG set all fields to 0).
+pub trait ErasePage<U> {
+    /// An enumeration of Storage errors
+    type Error;
+
+    /// Erase the page of memory
+    /// For flash devices, this sets the whole page to 0xFF
+    /// Implementations should mask the address as required to get the page to erase
+    fn try_erase_page(&mut self, page: Page<U>) -> nb::Result<(), Self::Error>;
+
+    /// Erase the page of memory at the address. Note: The only valid address is the start of the page
+    /// For flash devices, this sets the whole page to 0xFF
+    /// Implementations should mask the address as required to get the page to erase
+    fn try_erase_address(&mut self, address: Address<U>) -> nb::Result<(), Self::Error>;
+}
+
+/// Allow for checking that data can fit before writing to the device.
+///
+/// As some devices have limits on writing accross pages, the page size is also included.
+pub trait StorageSize<Word, U> {
+    /// An enumeration of Storage errors
+    type Error;
+
+    /// Returns the start address and the maximum size that can be stored by the device
+    fn try_total_size(&mut self) -> nb::Result<(Address<U>, AddressOffset<U>), Self::Error>;
```

Comment:
I think two different methods may be better, since it would be fairly easy to make a mistake.